### PR TITLE
More smoke and field balancing [RDY]

### LINKED ISSUE #17529: Vehicle smoke bugged
It looks like vehicles creates not light smoke but very dense fog, which stays for very long time. It's very annoying and tedious.
Last tested build was 5194.
It works well at least in build 5116.

### PR DESCRIPTION
**Pull request by mugling**
### Summary
- Smoke half life reduced (fixes #17529)
- Monsters JSON definitions can specify `emit_fields` as opposed to relying upon `special_attacks`
- Provides an `abstract` type for `grenade_canister` (8fff2a7)
- Riot control bot now fires `40mm_teargas` canisters
### Implementation

Allowing items and monsters to specify emission of fields from a limited set of `emit_id` defined in JSON is helpful for balancing. For example the smoker zombie and smoke bomb should have similar characteristics. This should be helpful as rework the `fields` code as each emission source only needs checking once to ensure no regressions.

Unlike `map::add_field` the new `map::emit_field` respects terrain obstacles as it internally uses @Coolthulhu's new function `map::propagate_field`. In particular gas grenades are now both much more realistic and have more tactical nuances than before. An abstract type `grenade_canister` is provided in `grenades.json` which may be especially useful for those producing mods.

This PR provides several examples of different uses, including flintlock `MUZZLE_SMOKE`, `mon_zombie_smoker` and especially `mon_riotbot`. The latter is now a much more capable opponent which adds some much needed variety to the midgame opponents. Tear gas is fairly dangerous stuff to the player but a gas mask or cover provide some mitigation.

### CODE REVIEW

**Comment by pisskop** [comment]
Does this invalidate the old special attacks?
Can bloated zombies do the same, and can it be other materials, such as teargas or fields of glowing plasma?

**Comment by mugling** [comment]
> Does this invalidate the old special attacks?

No, the smoker continues to try and bite and the riot control bot retains all it's previous tricks

> Can bloated zombies do the same, and can it be other materials, such as teargas or fields of glowing plasma?

Yes, any number of fields can emitted whilst the critter remains alive. Items continue to emit fields whilst they are `active`.

**Comment by Coolthulhu** [comment]
That giant gas nerf will pretty much make it not happen outside smokers and gas bombs. It would be a better idea to nerf it only halfway and then change the most powerful sources of it instead.

**Comment by mugling** [comment]
@Coolthulhu can you rephrase?

**Comment by Coolthulhu** [comment]
Look at the new half life of smoke. It pretty much guarantees it will drop in intensity each turn it spends on the outside and that it will disperse quickly even inside.

This means that the only kind of smoke that will matter will be the one recreated every turn - smokers and smoke bombs.

Also, unrelated note: I added a minor bug in the propagation PR. Smokers emit poison gas when dying, they emited smoke before.

**Comment by mugling** [comment]
> Look at the new half life of smoke. It pretty much guarantees it will drop in intensity each turn it spends on the outside and that it will disperse quickly even inside.

I buffed it a **lot** in #17331 from λ=180s to λ=600s along with dropping the outdoor penalty (which makes smoke defacto useless outdoors) and reducing dissipation from 80 to 30.

This PR restores the balance by reverting λ=120s and increasing dissipation to 50. I've checked smoker zombies, flintlock weapons, vehicle exhausts and fires all have reasonable output.

I'm starting to wonder if what I actually should have done is stick with λ=180s and go with dissipation 50 and no outdoor penalty although given the above examples all work correctly I'm reluctant to keep bouncing between the numbers.

I'm flexible either way but I want to try and keep this PR quick and small as once we can start using `emit_id` in more places balancing becomes much easier in future PR's.

**Comment by Coolthulhu** [comment]
Oh, I misread the value - I though you changed the outdoor dissipation time here, which is essentially a multiplier.

Still, original times were rather low and made smoke dissipate quickly.

Outdoor penalty for smoke should be a thing. It can be something like 4 or so, but it should exist.

Fires currently have very low smoke output, but that's more for performance than balance reasons.

**Comment by mugling** [comment]
> Oh, I misread the value

No worries. The field code in general is pretty nasty

> Outdoor penalty for smoke should be a thing. It can be something like 4 or so, but it should exist.

Any objection to leaving this PRas it is for now as it works. It's not perfect but I don't think we could guarantee overall balance in any case until we've addressed all the possible sources of `fd_smoke`. Once everything that can be `emit_id` is and other special cases identified we can consider that further?

> Fires currently have very low smoke output, but that's more for performance than balance reasons.

It's possible we need two sorts of smoke and certainly more granularity in density (even if we only display three types to the player). Most definitely a later PR.

**Comment by pisskop** [comment]
A nontoxic irritant and toxic version would be helpful.  For say a burning tree versus a burning deck with varnish on it.

**Comment by mugling** [comment]
Variants are more to do with persistance of the smoke

File: src/ranged.cpp [hunk 1]
```diff
@@ -192,12 +192,9 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
...
     }
 
+    static emit_id muzzle_smoke( "emit_smoke_plume" );
     if( proj_effects.count( "MUZZLE_SMOKE" ) ) {
-        for( const auto& e : closest_tripoints_first( 1, trajectory.empty() ? source : trajectory[ 0 ] ) ) {
-            if( one_in( 2 ) ) {
-                g->m.add_field( e, fd_smoke, 1, 0 );
-            }
-        }
+        g->m.emit_field( trajectory.empty() ? source : trajectory[ 0 ], muzzle_smoke );
```
Comment (codemime) [inline src/ranged.cpp]: The ternary won't be required if you move this line below [that one](https://github.com/CleverRaven/Cataclysm-DDA/pull/17530/files#diff-c5873ed460ae46963dac594829494e52R210). Also, `trajectory.front()` is cleaner.

**Comment by codemime** [comment]
Only "Smoky bear" and a couple of monsters from the PK mod also use the `SMOKECLOUD` flag. Can the flag be deprecated?

File: src/ranged.cpp (hunk 1 above)
Comment (mugling) [inline src/ranged.cpp]: Good point on both

**Comment by mugling** [comment]
Looking at the PK mod we badly need the ability for mods to partially override a `monster` definition

**Comment by mugling** [comment]
Resolved fairly challenging merge conflict via rebase and merge

**Comment by BevapDin** [comment]
> TANGENT: Looking at the PK mod we badly need the ability for mods to partially override a  monster  definition

What do you mean? This already possible:
```JSON
{
    "type" : "MONSTER",
    "id" : "debug_mon",
    "edit-mode": "modify",
    "name": "not the debug monster"
}
```

**Comment by mugling** [comment]
@BevapDin that I did not know

**Comment by pisskop** [comment]
I only overwrite the ones that evolve into specific critters instead of groups.

i.e. Scorched zombies evolve into_group: GROUP_ZOMBIE_UPGRADE whilst I upgrade them into mon_zom_scorched_upgrade.  The game gets upset that I add in my own upgrade as 'a duplicate' or redundant tag.



What would help is to have 'copy-from' apply to monsters as well as items.

**Comment by Coolthulhu** [comment]
Riot bots launch their gas canisters when trying to arrest. Tear gas is deadly.

Smoke needs an outdoors dispersal modifier - 2 minutes is OK for outdoors, but indoors it should be 10-20 or so.

**Comment by mugling** [comment]
> Riot bots launch their gas canisters when trying to arrest unarmed player. Tear gas is deadly.

Is this necessarily a problem? I quite want the riot bot to be a more deadly opponent?

**Comment by Coolthulhu** [comment]
The whole deal with riot bots is that they allow the player to choose not to die and instead get shackled.
Tear gas is at least an order of magnitude more powerful than poison gas. Riot bots shouldn't use it against unarmed players.

**Comment by Coolthulhu** [comment]
Oops, assigned without checking.

Should I wait for fixes?

**Comment by mugling** [comment]
I've reverted the riot bot changes.

I want to leave further smoke balancing until after this PR as it's probable, if not certain, that as we adjust more source of smoke to be produced via `emit` that further adjustments would be needed.

The current values close some existing bugs and are usable for all circumstances.

**Comment by mugling** [comment]
Ready

**Comment by Coolthulhu** [comment]
Smoke needs an outdoors dispersal modifier - 2 minutes is OK for outdoors, but indoors it should be 10-20 or so.

**Comment by mugling** [comment]
> I want to leave further smoke balancing until after this PR as it's probable, if not certain, that as we adjust more source of smoke to be produced via emit that further adjustments would be needed.
> 
> The current values close some existing bugs and are usable for all circumstances.
